✨ Add transfer feature for emissions-generating goo

BenSparksCode opened the pull request:
Gm sers.

Any opinions on adding a `transferGoo` function to the ArtGobblers contract, such as the one I've implemented here?

I think this would help composability by allowing users to transfer GOO to each other without going through the gas-costly process of withdrawing to ERC20 and then re-depositing each time.

I'll add a few tests for it as well if peeps are in favor 🫡

BenSparksCode commented:
Oof I just realized a `transferFrom` equivalent is actually what we need for composable applications. I'll implement that too if you think this idea has merit.

BenSparksCode commented on src/ArtGobblers.sol:
```diff
@@ -786,6 +794,76 @@ contract ArtGobblers is GobblersERC721, LogisticVRGDA, Owned, ERC1155TokenReceiv
+        getUserData[from].lastTimestamp = uint64(block.timestamp);
+        getUserData[to].lastTimestamp = uint64(block.timestamp);
+
+        emit GooBalanceUpdated(from, fromUpdatedBalance);
+        emit GooBalanceUpdated(to, toUpdatedBalance);
+        emit GooTransfer(from, to, gooAmount);
```
What's the stance on events here? Keep all three, or just use GooTransfer (which seems to convey all the same properties). On the other hand, supporting older events might be better for frontend integration.

OpenCoreCH commented:
Currently, when you do a transfer with the same source and destination, the balance of the destination gets decreased. This happens because `lastTimestamp` is only updated in the end, so `gooBalance(to)` will already use the new (decreased) balance, but still the old timestamp for the calculation. I do not have permission to add commits to the PR, but here is a commit (including a test that fails with the current implementation) that fixes this:

```diff
--- a/src/ArtGobblers.sol
+++ b/src/ArtGobblers.sol
@@ -813,6 +813,7 @@ contract ArtGobblers is GobblersERC721, LogisticVRGDA, Owned, ERC1155TokenReceiv
         // Will revert due to underflow if we're decreasing by more than the user's current balance.
         uint256 fromUpdatedBalance = gooBalance(msg.sender) - gooAmount;
         getUserData[msg.sender].lastBalance = uint128(fromUpdatedBalance);
+        getUserData[msg.sender].lastTimestamp = uint64(block.timestamp);
         // Don't need to do checked addition in the increase case.
         uint256 toUpdatedBalance;
         unchecked {
@@ -821,7 +822,6 @@ contract ArtGobblers is GobblersERC721, LogisticVRGDA, Owned, ERC1155TokenReceiv
         }
 
         // Update both accounts lastTimestamp now
-        getUserData[msg.sender].lastTimestamp = uint64(block.timestamp);
         getUserData[to].lastTimestamp = uint64(block.timestamp);
 
         emit GooBalanceUpdated(msg.sender, fromUpdatedBalance);
@@ -846,6 +846,7 @@ contract ArtGobblers is GobblersERC721, LogisticVRGDA, Owned, ERC1155TokenReceiv
         // Will revert due to underflow if we're decreasing by more than the user's current balance.
         uint256 fromUpdatedBalance = gooBalance(from) - gooAmount;
         getUserData[from].lastBalance = uint128(fromUpdatedBalance);
+        getUserData[from].lastTimestamp = uint64(block.timestamp);
         // Don't need to do checked addition in the increase case.
         uint256 toUpdatedBalance;
         unchecked {
@@ -854,7 +855,6 @@ contract ArtGobblers is GobblersERC721, LogisticVRGDA, Owned, ERC1155TokenReceiv
         }
 
         // Update both accounts lastTimestamp now
-        getUserData[from].lastTimestamp = uint64(block.timestamp);
         getUserData[to].lastTimestamp = uint64(block.timestamp);
 
         emit GooBalanceUpdated(from, fromUpdatedBalance);

--- a/test/ArtGobblers.t.sol
+++ b/test/ArtGobblers.t.sol
@@ -872,6 +872,18 @@ contract ArtGobblersTest is DSTestPlus {
         assertEq(gobblers.gooBalance(users[0]) - goo.balanceOf(users[0]), initialBalance - transferAmount);
     }
 
+    function testGooSelfTransfer() public {
+        mintGobblerToAddress(users[0], 1);
+        vm.warp(block.timestamp + 1 days);
+        setRandomnessAndReveal(1, "seed");
+        vm.warp(block.timestamp + 100000);
+        uint256 initialBalance = gobblers.gooBalance(users[0]);
+        uint256 transferAmount = initialBalance / 10; //10%
+        vm.prank(users[0]);
+        gobblers.transferGoo(users[0], transferAmount);
+        assertEq(gobblers.gooBalance(users[0]), initialBalance);
+    }
+
     /// @notice Test that an internal goo transfer reverts if attempting
     /// to transfer more goo than sender owns.
     function testCantTransferMoreGooThanOwned() public {
```

BenSparksCode commented:
> Currently, when you do a transfer with the same source and destination, the balance of the destination gets decreased. This happens because `lastTimestamp` is only updated in the end, so `gooBalance(to)` will already use the new (decreased) balance, but still the old timestamp for the calculation. I do not have permission to add commits to the PR, but here is a commit (including a test that fails with the current implementation) that fixes this:
> 
> ```diff
> --- a/src/ArtGobblers.sol
> +++ b/src/ArtGobblers.sol
> @@ -813,6 +813,7 @@ contract ArtGobblers is GobblersERC721, LogisticVRGDA, Owned, ERC1155TokenReceiv
>          // Will revert due to underflow if we're decreasing by more than the user's current balance.
>          uint256 fromUpdatedBalance = gooBalance(msg.sender) - gooAmount;
>          getUserData[msg.sender].lastBalance = uint128(fromUpdatedBalance);
> +        getUserData[msg.sender].lastTimestamp = uint64(block.timestamp);
>          // Don't need to do checked addition in the increase case.
>          uint256 toUpdatedBalance;
>          unchecked {
> @@ -821,7 +822,6 @@ contract ArtGobblers is GobblersERC721, LogisticVRGDA, Owned, ERC1155TokenReceiv
>          }
>  
>          // Update both accounts lastTimestamp now
> -        getUserData[msg.sender].lastTimestamp = uint64(block.timestamp);
>          getUserData[to].lastTimestamp = uint64(block.timestamp);
>  
>          emit GooBalanceUpdated(msg.sender, fromUpdatedBalance);
> @@ -846,6 +846,7 @@ contract ArtGobblers is GobblersERC721, LogisticVRGDA, Owned, ERC1155TokenReceiv
>          // Will revert due to underflow if we're decreasing by more than the user's current balance.
>          uint256 fromUpdatedBalance = gooBalance(from) - gooAmount;
>          getUserData[from].lastBalance = uint128(fromUpdatedBalance);
> +        getUserData[from].lastTimestamp = uint64(block.timestamp);
>          // Don't need to do checked addition in the increase case.
>          uint256 toUpdatedBalance;
>          unchecked {
> @@ -854,7 +855,6 @@ contract ArtGobblers is GobblersERC721, LogisticVRGDA, Owned, ERC1155TokenReceiv
>          }
>  
>          // Update both accounts lastTimestamp now
> -        getUserData[from].lastTimestamp = uint64(block.timestamp);
>          getUserData[to].lastTimestamp = uint64(block.timestamp);
>  
>          emit GooBalanceUpdated(from, fromUpdatedBalance);
> 
> --- a/test/ArtGobblers.t.sol
> +++ b/test/ArtGobblers.t.sol
> @@ -872,6 +872,18 @@ contract ArtGobblersTest is DSTestPlus {
>          assertEq(gobblers.gooBalance(users[0]) - goo.balanceOf(users[0]), initialBalance - transferAmount);
>      }
>  
> +    function testGooSelfTransfer() public {
> +        mintGobblerToAddress(users[0], 1);
> +        vm.warp(block.timestamp + 1 days);
> +        setRandomnessAndReveal(1, "seed");
> +        vm.warp(block.timestamp + 100000);
> +        uint256 initialBalance = gobblers.gooBalance(users[0]);
> +        uint256 transferAmount = initialBalance / 10; //10%
> +        vm.prank(users[0]);
> +        gobblers.transferGoo(users[0], transferAmount);
> +        assertEq(gobblers.gooBalance(users[0]), initialBalance);
> +    }
> +
>      /// @notice Test that an internal goo transfer reverts if attempting
>      /// to transfer more goo than sender owns.
>      function testCantTransferMoreGooThanOwned() public {
> ```

Amazing catch, nice! Added your test and contract code improvement, thanks.

I've also added the improvement and test to `transferGooFrom` and made it such that if `from` == `msg.sender` then it skips the approval logic. This would allow us to change `transferGoo` logic to simply calling `transferGooFrom` internally.

Imo this would be cleaner but very slightly less gas-efficient than the base `transferGoo` as it is now. Open to thoughts.

FrankieIsLost commented:
@BenSparksCode, thanks for the contribution. Unfortunately, after internal discussion, we decided not to merge. As you can see from @OpenCoreCH's comment, this is a pretty high risk area of the code. Given the proximity to launch, we think the risk outweighs the reward for including this set of functions. 

Thanks again, and apologies we weren't able to get this across the finish line.
